### PR TITLE
CVS-41742 GetOutputsInfo implementation for pipelines

### PR DESCRIPTION
**Pull request by dkalinowski**

### CODE REVIEW

File: src/pipelinedefinition.cpp [hunk 1]
```diff
@@ -384,4 +384,63 @@ Status PipelineDefinition::getInputsInfo(tensor_map_t& inputsInfo, ModelManager&
     return StatusCode::OK;
 }
 
+Status PipelineDefinition::getOutputsInfo(tensor_map_t& outputsInfo, const ModelManager& manager) const {
+    // Assumptions: this can only be called on available pipeline definition.
+    // Add check if available when pipeline status will be implemented.
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:389]: Side comment:
probably we should throw when PipelineStatus is not in AVAILABLE/AVAILABLE_LIKE state. However we do not have status yet.

File: src/test/ensemble_metadata_test.cpp [hunk 2]
```diff
@@ -120,6 +133,18 @@ TEST(EnsembleMetadata, MultipleNodesOnDifferentLevelsUsingTheSamePipelineInputs)
...
+    const auto& intermediate_result_from_sum = outputs.at("intermediate_result_from_sum");
+    EXPECT_EQ(intermediate_result_from_sum->getShape(), shape_t({1, SUM_MODEL_OUTPUT_SIZE}));
+    EXPECT_EQ(intermediate_result_from_sum->getPrecision(), InferenceEngine::Precision::FP32);
+
+    const auto& original_input_for_N2 = outputs.at("original_input_for_N2");
+    EXPECT_EQ(original_input_for_N2->getShape(), shape_t({}));
+    EXPECT_EQ(original_input_for_N2->getPrecision(), InferenceEngine::Precision::UNSPECIFIED);
 }
 
 TEST(EnsembleMetadata, EmptyPipelineReturnsCorrectInputInfo) {
```
Comment (atobiszei) [inline src/test/ensemble_metadata_test.cpp:150]: Rename

File: src/pipelinedefinition.cpp [hunk 3]
```diff
@@ -384,4 +384,63 @@ Status PipelineDefinition::getInputsInfo(tensor_map_t& inputsInfo, ModelManager&
...
+            }
+            case NodeKind::DL: {
+                auto instance = manager.findModelInstance(dependency_node_info->modelName, dependency_node_info->modelVersion.value_or(0));
+                if (!instance) {
+                    // TODO: Change to SPDLOG_DEBUG before release
+                    SPDLOG_INFO("Model:{} was unavailable during pipeline:{} outputs info fetching", dependency_node_info->modelName, this->getName());
+                    return StatusCode::MODEL_MISSING;
+                }
+                std::unique_ptr<ModelInstanceUnloadGuard> unloadGuard;
+                auto status = instance->waitForLoaded(0, unloadGuard);
```
Comment (atobiszei) [inline src/pipelinedefinition.cpp:421]: Maybe add test with mocked model instance which will perform long load (1s), and we will launch during that getInputs/outputs info and check that we do not wait for model to load? Eg check execution time with small threshold.

File: src/pipelinedefinition.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp:389]: Yes, when pipeline status is added, we can throw at the beginning of this function if it is not in available state. For the period of getInputsInfo/getOutputsInfo we shall keep unload guard and keep it available.

**Comment by dkalinowski** [comment]
TODO: Add test for getting metadata if one of models is not loaded yet.

File: src/test/ensemble_metadata_test.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/test/ensemble_metadata_test.cpp]: Done

File: src/pipelinedefinition.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/pipelinedefinition.cpp:421]: I have added test for calling GetInputsInfo/GetOutputsInfo when one of models is in LOADING state. Does it cover your suggestion?
https://github.com/openvinotoolkit/model_server/pull/337/files#diff-171cd48347cc255c38e835edff0458e1008d80517d18f922106abf9445074c6cR329

**Review by atobiszei — APPROVED** [review]

**Review by mzegla — APPROVED** [review]
Looks good. Can we unify naming before merge? 
We're mixing snake_case and camelCase variable names in getOutputsInfo method.

**Comment by dkalinowski** [comment]
> TODO: Add test for getting metadata if one of models is not loaded yet.



> Looks good. Can we unify naming before merge?
> We're mixing snake_case and camelCase variable names in getOutputsInfo method.

Done

**Pull request merged**